### ISSUE TITLE
XOF should be infinitely callable

### ISSUE THREAD

**Issue by GiacomoPope** [GiacomoPope/kyber-py#22]
At the moment we just ask for a large number of bytes which should work in matrix computations with the XOF.

We should instead use SHAKE as intended, but `hashlib` makes this annoying.

One solution would be to copy this code and use it here as well

https://github.com/GiacomoPope/dilithium-py/blob/main/shake_wrapper.py

**Comment by tomato42** [comment]
So, I'm guessing that python will update to allow XOF operation of the function at some point, but for now we need something better...

1. we can either do one call, but get enough data from it so that there's less than 1 in a million chance of hitting it
2. just implement SHAKE in python and have it there

I don't like https://github.com/GiacomoPope/dilithium-py/blob/main/shake_wrapper.py as that caches the whole input, which is not bounded, while output is bound in practice

**Comment by GiacomoPope** [comment]
So we currently have option 1 in place (i dont know if the probability is 1 in a million). Implementing python SHAKE will be slow (it's one of the Kyber bottlenecks) I would rather use pycryptodome's SHAKE which works as intended.

**Comment by tomato42** [comment]
how about having a fallback to a pure-python SHAKE if pycryptodome is not installed?

I really would prefer to keep the project pure-python; even at the cost of performance

**Comment by GiacomoPope** [comment]
in keeping with this, would you like to have a pure-python version of AES too then? As this is a dependency we need for the RNG in MLKEM

**Comment by tomato42** [comment]
but that is for tests only, not for interop, is it?

**Comment by GiacomoPope** [comment]
for Kyber it's just for tests but i believe for ML-KEM you're supposed to use this nist approved randomness source, so technically it needs to be in the class seeded with os randomness

**Comment by tomato42** [comment]
No, section 3.3 of FIPS 203-IPD states:

> These random bytes shall be generated using an approved RBG, as prescribed in NIST SP 800-90A, NIST SP 800-90B,
> and NIST SP 800-90C [15 , 16, 17]. Moreover, the RBG used shall have a security strength of at
> least 128 bits for ML-KEM-512 , at least 192 bits for ML-KEM-768 , and at least 256 bits for
> ML-KEM-1024

there is no requirement to use AES_128_CTR for that, in fact there's explicit requirement not to use it for anything stronger than ML-KEM-512

**Comment by tomato42** [comment]
and those RBG requirements are only for strict FIPS compliance (which we don't care for, it's not like we're going to FIPS 140-3 certify this implementation...), so using os randomness directly is fine

**Comment by GiacomoPope** [comment]
sorry, let me rephrase.

We cannot have ML-KEM with random values called from something like `os.urandom()` we need to seed some other deterministic randomness generator. Seeing as we have AES_128_CTR already, then using something like AES_256_CTR seems sensible as we need similar thing for the KAT generation. Did you have something else from NIST SP 800-90A, NIST SP 800-90B,
and NIST SP 800-90C in mind?

**Comment by GiacomoPope** [comment]
> So, I'm guessing that python will update to allow XOF operation of the function at some point, but for now we need something better...

> An API of this nature would make more sense to offer in a PyPI SHAKE specific package before we consider if such an API makes sense in the standard library. We want hashlib to focus on the most common very widely used standard things.

:(

https://github.com/python/cpython/issues/82198

**Comment by tomato42** [comment]
> we need to seed some other deterministic randomness generator. 

why?

**Comment by GiacomoPope** [comment]
I'm happy to just us `os.urandom()` then as default and allow `AES_256_CTR` as an option like we do for Kyber. 

For the XOF I think we should make a wrapper around `hashlib` to do more normal SHAKE like things.

**Comment by tomato42** [comment]
> For the XOF I think we should make a wrapper around hashlib to do more normal SHAKE like things.

probably... I'll do a statistical analysis how much we need to read from it later

**Comment by GiacomoPope** [comment]
I believe this analysis is already in several places. If I have time later I'll try and find a reference.

**Comment by GiacomoPope** [comment]
maybe we just do 5 blocks and ask for dinner if the code crashes:

https://cryptojedi.org/papers/terminate-20230516.pdf

![image](https://github.com/user-attachments/assets/c054cf5e-dc01-4cb8-a2c4-2718c1f6c3c6)
